File: src/algorithms/arrays/max-product-subarray/sources/max-product-subarray.py
# Max Product Subarray — O(n) tracking both max and min products to handle negative flips
def max_product_subarray(input_array: list[int]) -> dict:
    array_length = len(input_array)

    if array_length == 0:  # @step:initialize
        return {"max_product": 0, "start_index": 0, "end_index": 0}  # @step:initialize

    current_max = input_array[0]  # @step:initialize
    current_min = input_array[0]  # @step:initialize
    global_max = input_array[0]  # @step:initialize
    current_start = 0
    best_start = 0
    best_end = 0

    for scan_index in range(1, array_length):
        current_element = input_array[scan_index]  # @step:compare

        # When multiplying by a negative, max and min swap roles
        if current_element < 0:  # @step:compare
            current_max, current_min = current_min, current_max  # @step:compare

        # Extend or restart the subarray
        current_max = max(current_element, current_max * current_element)  # @step:compare
        current_min = min(current_element, current_min * current_element)  # @step:compare

        if current_max == current_element:  # @step:compare
            current_start = scan_index  # @step:compare

        if current_max > global_max:  # @step:compare
            global_max = current_max  # @step:compare
            best_start = current_start  # @step:compare
            best_end = scan_index  # @step:compare

    return {"max_product": global_max, "start_index": best_start, "end_index": best_end}  # @step:complete

File: src/algorithms/arrays/kadane-subarray/max-product-subarray/sources/MaxProductSubarray.cpp
// Max Product Subarray — O(n) tracking both max and min products to handle negative flips
#include <vector>
#include <algorithm>
#include <tuple>

std::tuple<int, int, int> maxProductSubarray(const std::vector<int>& inputArray) {
    int arrayLength = (int)inputArray.size();

    if (arrayLength == 0) {
        // @step:initialize
        return {0, 0, 0}; // @step:initialize
    }

    int currentMax = inputArray[0]; // @step:initialize
    int currentMin = inputArray[0]; // @step:initialize
    int globalMax = inputArray[0]; // @step:initialize
    int currentStart = 0;
    int bestStart = 0;
    int bestEnd = 0;

    for (int scanIndex = 1; scanIndex < arrayLength; scanIndex++) {
        int currentElement = inputArray[scanIndex]; // @step:compare

        // When multiplying by a negative, max and min swap roles
        if (currentElement < 0) { // @step:compare
            int tempMax = currentMax; // @step:compare
            currentMax = currentMin; // @step:compare
            currentMin = tempMax; // @step:compare
        }

        // Extend or restart the subarray
        currentMax = std::max(currentElement, currentMax * currentElement); // @step:compare
        currentMin = std::min(currentElement, currentMin * currentElement); // @step:compare

        if (currentMax == currentElement) { // @step:compare
            currentStart = scanIndex; // @step:compare
        }

        if (currentMax > globalMax) { // @step:compare
            globalMax = currentMax; // @step:compare
            bestStart = currentStart; // @step:compare
            bestEnd = scanIndex; // @step:compare
        }
    }

    return {globalMax, bestStart, bestEnd}; // @step:complete
}
